### ISSUE TITLE
Data structure in explore and the scribble usage

### ISSUE THREAD

**Issue by YINSONGHUI** [panzhiyi/CC4S#3]
Dear author,
1)why this happens, is the code itself wrong?  
File "D:\zPaper_pros\CC4S\CC4S-main\tool\imutils.py", line 208
    h, w = img.shash pe[0:2]

2)can i run it on windows11 with pycharm and anaconda?

3)what is exactly the variables in basic_function.py: 
dataset_path and dataset\ 
dataset\
train_path\ (in inference part)    why this is needed during inference time?

4)after pip install , there still are some packages uninstalled,such as in too/imutil.py

5)can your project extends to 3D data?

**Comment by yinsh01** [comment]
“All the computations are carried out on NVIDIA TITAN RTX GPUs.”
一块3060的话连推理都不能跑吗？是不是一块需要设置一下GPU的分配啊

**Comment by sun1233217T** [comment]
Thanks for your interest in our work! We make the following reply to the issues you raised:

1. We sincerely apologize for the mistakes in the project and we have updated the correct version.
2. We acknowledge that the code may encounter some issues when running on Windows 11. Please feel free to provide detailed descriptions of any problems you encounter. This project can be fully executed in a conda environment.
3. The `dataset_path` and `dataset` specify the file location and type of the dataset, respectively. The `train_path` in the evaluation is primarily used to generate pseudo-labels for the current training data, which are utilized in the second stage of training.
4. Please provide us with specific uninstalled packages and we will update our `requirements.txt`.
5. 3D data is not the target of our work, but we believe that the uncertainty and inconsistency faced by the scribble supervision also plague weakly supervised learning of 3D data.

**Comment by YINSONGHUI** [comment]
Thanks for your patient reply!

 1.I've re downloaded the new code, and there appears no wrongs from IDE any more.
 2.The packages are qualified after changing the code!
 3.About the third issue: If i just want to inference, is it a necessary step to generate the pseu-do labels through training?
   Can't  i just use the ckpt of your provided file?

  4.I’ve downloaded the offical pascal voc dataset, and how can i get the SegmentationAug/train.txt
![image](https://github.com/user-attachments/assets/3b4ccb79-7d80-47c8-871d-7698e686908a)

**Comment by sun1233217T** [comment]
Thank you for your suggestion. We have updated the code, and now you can use inference.sh to directly get the model result.
For the corresponding .txt file, please refer to [URSS]( https://github.com/panzhiyi/URSS/issues/1 ).

**Comment by YINSONGHUI** [comment]
Sorry that i still come across some problems:

  1. In inference, what is the "train_path" 
    self.parser.add_argument('train_path', default='None' ,metavar='TRAIN_PATH', help='path to the dataset training file')
  2.In this repo's data/voc2012/SegmentationAug, there are only 2 images,and during inference ,this happens:
![image](https://github.com/user-attachments/assets/27a0c9f9-1f69-4a38-a5ce-34a150454884)
  3.Can you share file organization in resource management including 3 kind of annotation ,voc data and Augmentation data? 
  4.And it seems that the 'contour' directory under VOC2012 hasn't been provided yet
![image](https://github.com/user-attachments/assets/1792eb49-5906-4e76-9cbf-eb494924ac0c)
  5.All the 'path_related' variables' directories seem not to be provided.
![image](https://github.com/user-attachments/assets/dd3485aa-076d-4d97-b356-e526f71cd90b)
6.Can you make a screen recording of the training steps and the file organization in Explorer?

**Comment by sun1233217T** [comment]
We sincerely apologize for not organizing the code properly. We make the following reply to the issues you raised:

1.The train_path in the evaluation is primarily used to generate pseudo-labels for the current training data, which is not utilized during the inference stage.

2.The file structure is as follows:

```plaintext ... 
VOC2012/
├── ImageSets/
│   ├── SegmentationAug/
│   │   ├── train.txt
├── pascal_2012_scribble/
│   ├── 2007_000032.png
│   ├── 2007_000033.png
├── JPEGImages/
│   ├── 2007_000032.jpg
│   ├── 2007_000033.jpg
├── superpixels/
│   ├── 2007_000032.png
│   ├── 2007_000033.png
├── spiex/
│   ├── 2007_000032.mat
│   ├── 2007_000033.mat
├── Laplace/
│   ├── 2007_000032.mat
│   ├── 2007_000033.mat
```

The `spiex` and `Laplace` folders should be generated using `tool/scribblesup.m`, while the `superpixels` folder is available at: [here](https://drive.google.com/drive/folders/1q2PvbQVOdIY9S-qjh85ohM66svzp9wnp). The `counter` folder is actually not required; in the code, it generally refers to the `superpixels` folder.

**Comment by YINSONGHUI** [comment]
Thanks a lot!

  1. The pascal_2012_sribble folder contains xxx.xml files, so i need to transfer it to png?
  2. And the other VOC2012/SegmentationClassAug only contains 2 images, doesn't it corespond to ImageSets/SegmentationAug/train.txt?

**Comment by sun1233217T** [comment]
Thanks for your questions!
1.Yes, the .xml files should be converted to .png format. You can use the tool provided [here](https://github.com/meng-tang/rloss/blob/master/data/pascal_scribble/convertscribbles.m). The scribble files should be placed in the `pascal_2012_scribble` folder for correct usage.
2.The PASCAL VOC 2012 dataset contains 12,031 images, which are stored in the `JPEGImages` folder. You can directly download the corresponding `VOC2012/SegmentationClassAug` from [this link](https://www.dropbox.com/s/oeu149j8qtbs1x0/SegmentationClassAug.zip).

**Comment by YINSONGHUI** [comment]
Thanks for your considerately reply!

  1. I've transformed the .xml files in pascal_2012 folder to .png files and placed these pngs at data/VOC2012/pascal_2012_scribble 
  folder. Does this folder corespond to train_path in both training and evaluation .sh file?
![image](https://github.com/user-attachments/assets/cd0bda4c-d193-413b-b23c-e47b4a5fe779)

  2. Why there's minicoco under data/VOC2012. I saw it in scribblesup.m file
![image](https://github.com/user-attachments/assets/d2e3c0c0-22cf-4b95-ab2f-4fe33bcad6d4)

  3. The prediction of your provided model are all black, and i check the pixel value,finding out that some of the pixels have values and the picture shows very slight grey contour. Should i use the palette to manually transform them for better viualization?
![2007_000032](https://github.com/user-attachments/assets/81d5a46c-e4f3-4196-95fd-a5685b72288d)

  4.These two test() functions generate 10582(train.txt) plus 1449(val.txt) images, and what is the result?

      Are the results the pesudo labels for phase 2 training or the inference of the model itself cause i used evaluate.py with ckpt 
      xxx//CC4S_resnet50/model_best.pth.tar.

      If the results are pesudo labels for phase 2 training, then "scribble_path in dataset like:pascal_scribble_2012 \"in different .sh files means what?  What the diffenence between the pascal_scribble_2012 folders of transformed .xml files and the results of my problem 4 mentioned above?

![image](https://github.com/user-attachments/assets/7efc75e4-f1b4-4b47-a695-773325cf8e16)

  5.And the gt in train_seg_baseline.sh coresponds to what folder?

Thanks for all your patience! I've been helped  a lot!

**Comment by YINSONGHUI** [comment]
It suddently strikes me that when use your model ,what input should user provide?

For better explaination what i mean:
   In normal segmentation ,I input a image and the model output the segmentation result image.

   In your model, i shell input image and scribble label in order to get the segmentation result?

**Comment by sun1233217T** [comment]
1. Yes, by default, we assume that the scribble annotations are placed in the `pascal_2012_scribble `folder, and this parameter is set in the various shell scripts.
2. We tried testing on the COCO dataset.
3. The pure black output is for the convenience of code storage; if visualization is needed, you can use `inference.sh.`
4. When generating pseudo-labels, the train and test parts of the dataset are processed separately (though in practice, only the train part is needed).
5.Our ground truth (GT) is the scribble annotations stored in the `pascal_2012_scribble `folder.

In a semantic segmentation task, the user inputs an image and obtains a segmentation output. To train such a neural network, a set of images and corresponding supervision labels (fully annotated, scribble-annotated, or others) is required. Once training is complete, the neural network gains the ability to understand and segment other arbitrary images without any label.

**Comment by YINSONGHUI** [comment]
Thanks for answering all my questions!

  1. Please tell me how can i get the kind of input image like SegmentationClassAug for my own dataset.They are black-white style different from PASCAL VOC's offical SegmentationClass.
  And if i want to label my own dataset, how can i transform the label form:I manually label on the JEPGImages-like image resulting a mixed label-and-image style,and how can it be transformed to the one only containing label?

![image](https://github.com/user-attachments/assets/4fecd34d-efb3-446f-adce-1bb03c37f442)
TO:    
![image](https://github.com/user-attachments/assets/ea8fa20e-d772-4b0e-9830-b90b820de69f)

**Comment by YINSONGHUI** [comment]
1)I've read the code and changed the third-to-last parameter from" False" to "1" in order to make  if_mask==1, and the result is no longer pure black or white.

2)I figure that the "path" in evaluate.sh or inference.sh may be used to compute accuracy, but whether it is a litter bit confusing to request that parameter if the user just want to use the trained model to generate a segmentation result like the fourth picture above.

Please forgive me if i misunderstand whatever something.

You and Pan are very kind and professional cause i see both of you are warmhearted and all the reply are decent and formal,and i admire that.
